pkg/app/ops/insightcollector/collector.go

```diff
@@ -267,160 +282,162 @@ func (i *InsightCollector) getDailyInsightData(
 	var updatedps []insight.DataPoint
 
 	to := movePoint(rangeFrom, 1)
-	until := movePoint(rangeTo, 1)
-	var accumulatedTo time.Time
-	for {
+	for !to.After(rangeTo) {
 		targetTimestamp := insight.NormalizeTime(rangeFrom, step).Unix()
 
 		var data insight.DataPoint
-		var a time.Time
-		var err error
 		switch kind {
 		case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
-			data, a, err = i.getInsightDataForDeployFrequency(ctx, projectID, appID, targetTimestamp, rangeFrom, to)
+			data, deployments = i.extractDeployFrequency(deployments, rangeFrom.Unix(), to.Unix(), targetTimestamp)
 		case model.InsightMetricsKind_CHANGE_FAILURE_RATE:
-			data, a, err = i.getInsightDataForChangeFailureRate(ctx, projectID, appID, targetTimestamp, rangeFrom, to)
+			data, deployments = i.extractChangeFailureRate(deployments, rangeFrom.Unix(), to.Unix(), targetTimestamp)
 		default:
-			return nil, 0, fmt.Errorf("invalid step: %v", kind)
-		}
-		if err != nil {
-			if err == ErrDeploymentNotFound {
-				if to.Equal(until) {
-					break
-				}
-				rangeFrom = to
-				to = movePoint(to, 1)
-				continue
-			}
-			return nil, 0, err
+			return nil, fmt.Errorf("invalid step: %v", kind)
 		}
 
 		updatedps = append(updatedps, data)
-		rangeFrom = a
-		accumulatedTo = a
+		rangeFrom = movePoint(rangeFrom, 1)
+		to = movePoint(to, 1)
 	}
 
-	return updatedps, accumulatedTo.Unix(), nil
+	return updatedps, nil
 }
 
-var (
-	ErrDeploymentNotFound = errors.New("deployments not found")
-)
-
-// getInsightDataForDeployFrequency accumulate insight data in target range for deploy frequency.
-func (i *InsightCollector) getInsightDataForDeployFrequency(
+func (i *InsightCollector) getDeploymentsWithCreatedAt(
 	ctx context.Context,
-	projectID, applicationID string,
-	targetTimestamp int64,
-	from time.Time,
-	to time.Time) (*insight.DeployFrequency, time.Time, error) {
+	from, to int64,
+) ([]*model.Deployment, error) {
+
 	filters := []datastore.ListFilter{
 		{
 			Field:    "CreatedAt",
 			Operator: ">=",
-			Value:    from.Unix(),
-		},
-		{
-			Field:    "CreatedAt",
-			Operator: "<",
-			Value:    to.Unix(),
+			Value:    from,
 		},
 	}
 
-	if applicationID != "" {
-		filters = append(filters, datastore.ListFilter{
-			Field:    "ApplicationId",
-			Operator: "==",
-			Value:    applicationID,
-		})
-	}
-
-	if projectID != "" {
-		filters = append(filters, datastore.ListFilter{
-			Field:    "ProjectId",
-			Operator: "==",
-			Value:    projectID,
+	var deployments []*model.Deployment
+	maxCreatedAt := to
+	for {
+		d, err := i.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
+			PageSize: pageSize,
+			Filters: append(filters, datastore.ListFilter{
+				Field:    "CreatedAt",
+				Operator: "<",
+				Value:    maxCreatedAt,
+			}),
+			Orders: []datastore.Order{
+				{
+					Field:     "CreatedAt",
+					Direction: datastore.Desc,
+				},
+			},
```

Comment:
I may be wrong in some of my thinking... 🤔 
Call deployments in the range [from, maxCreatedAt) and update maxCreatedAt with minimal createdAt in first call's deployments. So, the response from continuous calls are not containing duplicate deployments, I think.

Anyway, I'll change it since some people, including you, feel it's simpler.

Reply:
Nice catch! You are right. I thought in reverse order.
Thank you.